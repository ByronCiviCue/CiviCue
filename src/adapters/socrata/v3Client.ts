import { socrataHeadersFor, socrataV3AuthHeader } from '../../lib/env-providers/socrata.js';
import { buildSocrataUrl, socrataFetch, SocrataHttpError } from '../../lib/http/socrata.js';
import { createClientError, isSocrataClientError } from './types.js';

export interface V3PostQueryInput {
  domain: string;
  datasetId: string; // abcd-1234
  soql: string; // full SoQL string
  pageNumber?: number; // default 1
  pageSize?: number; // default 1000
  includeSynthetic?: boolean; // default true
}

export interface V3PostQueryResult<T = unknown> {
  rows: T[];
  nextPageNumber?: number; // undefined when no more pages
  raw?: unknown;
}

export function isV3Unavailable(e: unknown): boolean {
  if (isSocrataClientError(e)) {
    const s = (e as { error: { status?: number } }).error.status;
    return s === 401 || s === 403 || s === 404 || s === 501;
  }
  if (e instanceof SocrataHttpError) {
    return e.status === 401 || e.status === 403 || e.status === 404 || e.status === 501;
  }
  return false;
}

export async function v3PostQuery<T = unknown>(input: V3PostQueryInput, signal?: AbortSignal): Promise<V3PostQueryResult<T>> {
  const { domain, datasetId } = input;
  const path = `/api/v3/views/${datasetId}/query.json`;
  const url = buildSocrataUrl(domain, path);
  const headers: Record<string, string> = {
    Accept: 'application/json',
    'Content-Type': 'application/json',
    ...socrataHeadersFor(domain),
    ...socrataV3AuthHeader(domain, datasetId)
  };
  const size = Math.min(Math.max(input.pageSize ?? 1000, 1), 1000);
  const body = JSON.stringify({
    query: input.soql,
    page: { pageNumber: input.pageNumber ?? 1, pageSize: size },
    includeSynthetic: input.includeSynthetic ?? true
  });
  const res = await socrataFetch(url.toString(), { method: 'POST', headers, body, signal });
  if (!res.ok) {
    // Never include auth header in error; redact
    const safeUrl = res.url;
    throw createClientError('HttpError', { status: res.status, url: safeUrl, message: `HTTP error (${res.status}) for ${safeUrl}` });
  }
  const json: unknown = await res.json();
  const maybeObj = (json && typeof json === 'object') ? (json as Record<string, unknown>) : undefined;
  let rows: unknown[] = [];
  if (maybeObj && Array.isArray((maybeObj as { data?: unknown }).data)) {
    rows = ((maybeObj as { data?: unknown }).data as unknown[]) ?? [];
  } else if (Array.isArray(json as unknown[])) {
    rows = json as unknown[];
  }
  const nextPageNumber = rows.length === size ? (input.pageNumber ?? 1) + 1 : undefined;
  return { rows: rows as T[], nextPageNumber, raw: json };
}

export async function v3PostAll<T = unknown>(
  domain: string,
  datasetId: string,
  req: { query: string; includeSynthetic?: boolean; pageSize?: number },
  signal?: AbortSignal
): Promise<T[]> {
  const size = Math.min(Math.max(req.pageSize ?? 1000, 1), 1000);
  const out: T[] = [];
  let pageNumber = 1;
  // Loop until we receive a short page
  while (true) {
    const { rows } = await v3PostQuery<T>({ domain, datasetId, soql: req.query, pageNumber, pageSize: size, includeSynthetic: req.includeSynthetic }, signal);
    out.push(...rows);
    if (rows.length < size) break;
    pageNumber += 1;
  }
  return out;
}
