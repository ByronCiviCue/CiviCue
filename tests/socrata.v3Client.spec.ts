import { describe, it, expect, vi } from 'vitest';
import { v3PostQuery, isV3Unavailable } from '../src/adapters/socrata/v3Client.js';
import { withCassette } from './helpers/httpCassette.js';

vi.mock('../src/lib/env-providers/socrata.js', async () => {
  const actual = await vi.importActual<any>('../src/lib/env-providers/socrata.js');
  return {
    ...actual,
    socrataHeadersFor: (host: string) => (host.includes('token') ? { 'X-App-Token': 'tok' } : {}),
  };
});

describe('v3PostQuery', () => {
  const domain = 'data.sfgov.org';
  const datasetId = 'abcd-1234';

  it('posts with proper body and headers; no auth when env missing', async () => {
    const fetch = vi.spyOn(globalThis, 'fetch' as any).mockResolvedValue(new Response(JSON.stringify({ data: [{ id: 1 }] }), { status: 200, headers: { 'content-type': 'application/json' } }));
    const res = await v3PostQuery({ domain, datasetId, soql: 'select id', pageNumber: 1, pageSize: 1 });
    expect(res.rows).toEqual([{ id: 1 }]);
    const [calledUrl, init] = fetch.mock.calls[0] as [string, RequestInit];
    expect(calledUrl).toContain(`/api/v3/views/${datasetId}/query.json`);
    expect(init.method).toBe('POST');
    const body = JSON.parse(init.body as string);
    expect(body).toMatchObject({ query: 'select id', page: { pageNumber: 1, pageSize: 1 }, includeSynthetic: true });
    expect((init.headers as any)['Authorization']).toBeUndefined();
    fetch.mockRestore();
  });

  it('adds Basic auth when env provides keys', async () => {
    // Patch env for this test
    (process.env as any)[`SOCRATA__${domain}__V3_KEY_ID`] = 'kid';
    (process.env as any)[`SOCRATA__${domain}__V3_KEY_SECRET`] = 'ksec';
    const fetch = vi.spyOn(globalThis, 'fetch' as any).mockResolvedValue(new Response(JSON.stringify({ data: [] }), { status: 200 }));
    await v3PostQuery({ domain, datasetId, soql: 'select *', pageNumber: 1, pageSize: 1 });
    const init = fetch.mock.calls[0][1] as RequestInit;
    const auth = (init.headers as any)['Authorization'];
    expect(auth).toMatch(/^Basic /);
    fetch.mockRestore();
    delete (process.env as any)[`SOCRATA__${domain}__V3_KEY_ID`];
    delete (process.env as any)[`SOCRATA__${domain}__V3_KEY_SECRET`];
  });

  it('computes nextPageNumber when page is full', async () => {
    const fetch = vi.spyOn(globalThis, 'fetch' as any).mockResolvedValue(new Response(JSON.stringify({ data: [{}, {}] }), { status: 200 }));
    const r = await v3PostQuery({ domain, datasetId, soql: 'select *', pageNumber: 2, pageSize: 2 });
    expect(r.nextPageNumber).toBe(3);
    fetch.mockRestore();
  });

  it('v3PostAll paginates until short page and concatenates', async () => {
    const pages = [
      { data: [{ id: 1 }, { id: 2 }] },
      { data: [{ id: 3 }, { id: 4 }] },
      { data: [{ id: 5 }] },
    ];
    let call = 0;
    const fetch = vi.spyOn(globalThis, 'fetch' as any).mockImplementation(async () => {
      const body = JSON.stringify(pages[Math.min(call, pages.length - 1)]);
      call++;
      return new Response(body, { status: 200, headers: { 'content-type': 'application/json' } });
    });
    const all = await (await import('../src/adapters/socrata/v3Client.js')).v3PostAll(domain, datasetId, { query: 'select id', pageSize: 2 });
    expect(all).toHaveLength(5);
    expect(fetch).toHaveBeenCalledTimes(3);
    fetch.mockRestore();
  });

  it('clamps pageSize to 1..1000 in request body', async () => {
    const fetch = vi
      .spyOn(globalThis, 'fetch' as any)
      .mockResolvedValueOnce(new Response(JSON.stringify({ data: [] }), { status: 200 }))
      .mockResolvedValueOnce(new Response(JSON.stringify({ data: [] }), { status: 200 }));

    // pageSize above max clamps to 1000
    await v3PostQuery({ domain, datasetId, soql: 'select *', pageSize: 5000 });
    let sent = JSON.parse((fetch.mock.calls[0][1] as RequestInit).body as string);
    expect(sent.page.pageSize).toBe(1000);

    // pageSize below min clamps to 1
    await v3PostQuery({ domain, datasetId, soql: 'select *', pageSize: 0 });
    sent = JSON.parse((fetch.mock.calls[1][1] as RequestInit).body as string);
    expect(sent.page.pageSize).toBe(1);

    fetch.mockRestore();
  });

  it('supports AbortSignal and yields AbortError without partial results', async () => {
    const controller = new AbortController();
    vi.stubGlobal('fetch', ((_: string, init: any) => {
      const sig = init.signal as AbortSignal | undefined;
      if (sig?.aborted) {
        const err: any = new Error('Aborted');
        err.name = 'AbortError';
        return Promise.reject(err);
      }
      // Never resolve to simulate long request; caller will abort
      return new Promise((_resolve: any, reject: any) => {
        sig?.addEventListener('abort', () => {
          const err: any = new Error('Aborted');
          err.name = 'AbortError';
          reject(err);
        });
      });
    }) as any);
    const p = v3PostQuery({ domain, datasetId, soql: 'select *' }, controller.signal);
    controller.abort();
    await expect(p).rejects.toHaveProperty('name', 'AbortError');
    vi.unstubAllGlobals();
  });

  it('never leaks secrets in errors', async () => {
    const kid = 'myKeyId';
    const ksec = 'mySecret';
    (process.env as any)[`SOCRATA__${domain}__V3_KEY_ID`] = kid;
    (process.env as any)[`SOCRATA__${domain}__V3_KEY_SECRET`] = ksec;
    const fetch = vi.spyOn(globalThis, 'fetch' as any).mockResolvedValue(new Response('Forbidden', { status: 403 }));
    const p = v3PostQuery({ domain, datasetId, soql: 'select *' });
    await expect(p).rejects.toMatchObject({ error: { kind: 'HttpError', status: 403 } });
    try { await p; } catch (e: any) {
      const s = JSON.stringify(e);
      const basic = Buffer.from(`${kid}:${ksec}`).toString('base64');
      expect(s).not.toContain(kid);
      expect(s).not.toContain(ksec);
      expect(s).not.toContain(basic);
    }
    fetch.mockRestore();
    delete (process.env as any)[`SOCRATA__${domain}__V3_KEY_ID`];
    delete (process.env as any)[`SOCRATA__${domain}__V3_KEY_SECRET`];
  });

  it('flags v3 unavailable errors for graceful fallback', async () => {
    const fetch = vi.spyOn(globalThis, 'fetch' as any).mockResolvedValue(new Response('Forbidden', { status: 403 }));
    const p = v3PostQuery({ domain, datasetId, soql: 'select *' });
    await expect(p).rejects.toMatchObject({ error: { kind: 'HttpError', status: 403 } });
    try { await p; } catch (e) { expect(isV3Unavailable(e)).toBe(true); }
    fetch.mockRestore();
  });

  it('replays a basic v3 POST page via cassette', async () => {
    const result = await withCassette('v3-client-basic', async () => {
      const r = await v3PostQuery({ domain: 'data.cassette.test', datasetId: 'abcd-1234', soql: 'select id' });
      return r.rows;
    }, { mode: 'replay' });
    expect(Array.isArray(result)).toBe(true);
    expect(result.length).toBeGreaterThanOrEqual(3);
    for (const row of result as any[]) {
      expect(row).toHaveProperty('id');
    }
    // Stable-field checks to avoid snapshot brittleness
    const ids = (result as any[]).map(r => r.id);
    expect(ids).toEqual([1, 2, 3]);
  });
});
