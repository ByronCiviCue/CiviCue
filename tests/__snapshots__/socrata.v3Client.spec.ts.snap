// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`v3PostQuery > replays a basic v3 POST page via cassette 1`] = `
[
  {
    "id": 1,
  },
  {
    "id": 2,
  },
  {
    "id": 3,
  },
]
`;
